src/models/booking.cr

```diff
@@ -182,6 +242,84 @@ class Booking
     where("( #{query} )")
   end
 
+  # Booking starts in the future, no one has checked-in and it hasn't been deleted
+  protected def is_reserved?(current_time : Int64 = Time.local.to_unix)
+    booking_start > current_time &&
+      !checked_in_column.value(false) &&
+      !deleted_column.value(false) &&
+      !rejected_column.value(false)
+  end
+
+  # Booking is currently active (the wall clock time is between start and end times of the booking) and the user has checked in
+  protected def is_checked_in?(current_time : Int64 = Time.local.to_unix)
+    checked_in &&
+      booking_start <= current_time &&
+      booking_end >= current_time
+  end
+
+  # The user checked out during the start and end times
+  protected def is_checked_out?
+    (co_at = checked_out_at_column.value(nil)) &&
+      booking_start <= co_at &&
+      booking_end >= co_at
+  end
+
+  # It's past the end time of the booking and it was never checked in
+  protected def is_no_show?(current_time : Int64 = Time.local.to_unix)
+    !checked_in_at_column.value(nil) &&
+      booking_end < current_time
+  end
+
+  # Someone rejected the booking before it started
+  protected def is_rejected?
+    rejected &&
+      (r_at = rejected_at) &&
+      booking_start > r_at
+  end
+
+  # The booking was deleted before the booking start time
+  protected def is_cancelled?
+    deleted_column.value(false) &&
+      (del_at = deleted_at) &&
+      booking_start > del_at
+  end
+
+  # The current time is past the end of the booking, the user checked-in but never checked-out
+  protected def is_ended?(current_time : Int64 = Time.local.to_unix)
+    !checked_out_at_column.value(nil) &&
+      checked_in &&
+      booking_end < current_time
+  end
+
+  def current_state : State
+    current_time = Time.local.to_unix
+
+    case self
+    when .is_reserved?(current_time)   then State::Reserved
+    when .is_checked_in?(current_time) then State::CheckedIn
+    when .is_checked_out?              then State::CheckedOut
+    when .is_no_show?(current_time)    then State::NoShow
+    when .is_rejected?                 then State::Rejected
+    when .is_cancelled?                then State::Cancelled
+    when .is_ended?                    then State::Ended
+    else
+      unknown_state = {
+        current_time: current_time,
+        booking_start: booking_start,
+        booking_end: booking_end,
+        rejected: rejected,
+        rejected_at: rejected_at_column.value(nil),
+        checked_in: checked_in,
+        checked_in_at: checked_in_at_column.value(nil),
+        checked_out_at: checked_out_at_column.value(nil),
+        deleted: deleted_column.value(nil),
+        deleted_at: deleted_at_column.value(nil),
+      }.to_json
+      Log.error { "Booking is in an Unknown state: #{unknown_state}" }
+      State::Unknown
```

Comment:
@caspiano I have moded the error log to here.

Thinking about it, it might be possible to get rid of some of the Unknown outcomes if I changed the checks to only use the _at versions instead of rejected, checked_in, deleted.

Reply:
Good idea. Those non-`_at` fields strike me as slightly redundant.

Reply:
I changed it to use the _at fields instead.

Reply:
I changed it to use the _at fields instead.

Reply:
Missing the `booking_id` in the log here.
Not sure if this is set up in `staff-api` but I hacked an extension of `Log` together that allows you to pass data via namedtuple/hash

https://github.com/spider-gazelle/log_helper

Reply:
id added.